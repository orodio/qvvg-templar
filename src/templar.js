export const interleave = (a = [], b = [], c = []) => {
  if (!a.length && !b.length) return c
  if (!a.length) return c
  if (!b.length) return [...c, a[0]]

  const [aHead, ...aRest] = a
  const [bHead, ...bRest] = b

  if (aHead !== undefined) c.push(aHead)
  if (bHead !== undefined) c.push(bHead)

  return interleave(aRest, bRest, c)
}

const recApply = d => arg1 => {
  if (typeof arg1 === "function") return recApply(d)(arg1(d))
  if (typeof arg1 === "string") return arg1
  return ""
}

export const t7l = (head, ...rest) => {
  if (typeof head === "string") return () => head
  if (Array.isArray(head)) {
    return d =>
      interleave(head, rest.map(recApply(d)))
        .join("")
        .trim()
  }
  return head
}
