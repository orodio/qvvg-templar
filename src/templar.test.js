import {t7l, interleave} from "./templar"

describe("interleave", () => {
  ;[
    [[], [], []],
    [[1], [], [1]],
    [[], [1], []],
    [[1], [2], [1, 2]],
    [
      [1, 2, 3],
      [4, 5, 6],
      [1, 4, 2, 5, 3, 6],
    ],
    [
      [1, 2, 3],
      [4, 5],
      [1, 4, 2, 5, 3],
    ],
  ].forEach(([a, b, c]) => {
    test(`interleave([${a}], [${b}]) -> [${c}]`, () => {
      expect(interleave(a, b)).toStrictEqual(c)
    })
  })
})

const _ = (msg, a, b) => {
  if (b == null) {
    test(msg, () => expect(a).toMatchSnapshot())
  } else {
    test(msg, () => expect(a).toBe(b))
  }
}
const t = v => typeof v

describe("t7l", () => {
  describe("input type vs output type", () => {
    _("t7l`` -> function", t(t7l``))
    _("t71(``) -> function", t(t7l("words")))
    _("t7l(t7l``) -> function", t(t7l(t7l``)))
  })

  describe("no interop", () => {
    _("t7l`abc`() -> 'abc'", t7l`abc`())
    _("t7l('abc')() -> 'abc'", t7l("abc")())
    _("t7l(t7l`abc`)() -> 'abc'", t7l(t7l`abc`)())
  })

  describe("only interop", () => {
    _("t7l`${'abc'}`() -> 'abc'", t7l`${"abc"}`())
    _("t7l(t7l`${'abc'}`)() -> 'abc'", t7l(t7l`${"abc"}`)())
  })

  describe("interop function", () => {
    const o = {a: "abc"}
    _("t7l`${o=>o.a}`(o) -> 'abc'", t7l`${o => o.a}`(o))
    _("t7l(t7l`${o=>o.a}`)(o) -> 'abc'", t7l(t7l`${o => o.a}`)(o))
  })

  describe("interop more t7l", () => {
    const o = {a: "abc"}
    _("t7l`x${t7l`y${o=>o.a}`}`(o) => 'xyabc'", t7l`x${t7l`y${o => o.a}`}`(o))
    _(
      "t7l`x${t7l`y${t7l`z${o=>o.a}`}`}`(o) => 'xyabc'",
      t7l`x${t7l`y${t7l`z${o => o.a}`}`}`(o)
    )
    _(
      "t7l(t7l`x${t7l`y${o => o.a}`}`)(o) -> 'xyabc'",
      t7l(t7l`x${t7l`y${o => o.a}`}`)(o)
    )
  })

  describe("interop nested functions", () => {
    const fn = a => b => c => d => e => f => f.a
    const o = {a: "abc"}
    _("t7l`${fn}`(o) -> 'abc'", t7l`${fn}`(o))
  })
})
