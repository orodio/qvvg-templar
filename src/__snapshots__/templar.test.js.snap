// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`t7l input type vs output type t7l(t7l\`\`) -> function 1`] = `"function"`;

exports[`t7l input type vs output type t7l\`\` -> function 1`] = `"function"`;

exports[`t7l input type vs output type t71(\`\`) -> function 1`] = `"function"`;

exports[`t7l interop function t7l(t7l\`\${o=>o.a}\`)(o) -> 'abc' 1`] = `"abc"`;

exports[`t7l interop function t7l\`\${o=>o.a}\`(o) -> 'abc' 1`] = `"abc"`;

exports[`t7l interop more t7l t7l(t7l\`x\${t7l\`y\${o => o.a}\`}\`)(o) -> 'xyabc' 1`] = `"xyabc"`;

exports[`t7l interop more t7l t7l\`x\${t7l\`y\${o=>o.a}\`}\`(o) => 'xyabc' 1`] = `"xyabc"`;

exports[`t7l interop more t7l t7l\`x\${t7l\`y\${t7l\`z\${o=>o.a}\`}\`}\`(o) => 'xyabc' 1`] = `"xyzabc"`;

exports[`t7l interop nested functions t7l\`\${fn}\`(o) -> 'abc' 1`] = `"abc"`;

exports[`t7l no interop t7l('abc')() -> 'abc' 1`] = `"abc"`;

exports[`t7l no interop t7l(t7l\`abc\`)() -> 'abc' 1`] = `"abc"`;

exports[`t7l no interop t7l\`abc\`() -> 'abc' 1`] = `"abc"`;

exports[`t7l only interop t7l(t7l\`\${'abc'}\`)() -> 'abc' 1`] = `"abc"`;

exports[`t7l only interop t7l\`\${'abc'}\`() -> 'abc' 1`] = `"abc"`;
